crates/fluvio-protocol/src/core/bytebuf.rs

```diff
@@ -0,0 +1,301 @@
+
+        assert!(bytebuf.is_empty());
+        assert!(bytebuf.inner.is_empty());
+    }
+
+    #[test]
+    fn test_is_empty_from_inner_vec() {
+        let bytebuf = ByteBuf::from(vec![128, 129, 130, 131]);
+
+        assert_eq!(bytebuf.is_empty(), bytebuf.inner.is_empty());
+    }
+
+    #[test]
+    fn test_is_empty_from_inner_empty_vec() {
+        let bytebuf = ByteBuf::from(Vec::default());
+
+        assert_eq!(bytebuf.is_empty(), bytebuf.inner.is_empty());
+    }
+
+    #[test]
+    fn test_len_from_inner_vec() {
+        let bytebuf = ByteBuf::from(vec![128, 129, 130, 131]);
+
+        assert_eq!(bytebuf.len(), bytebuf.inner.len());
+    }
+
+    #[test]
+    fn test_encode_bytebuf() {
+        let mut dest = Vec::default();
+        let value: ByteBuf = ByteBuf::from(vec![0x10, 0x11]);
+        let result = value.encode(&mut dest, 0);
+
+        assert!(result.is_ok());
+
+        // Length + Contents
+        assert_eq!(dest.len(), 6);
+
+        // Length Bytes
+        assert_eq!(dest[0], 0x00);
+        assert_eq!(dest[1], 0x00);
+        assert_eq!(dest[2], 0x00);
+        assert_eq!(dest[3], 0x02);
+
+        // Actual Content
+        assert_eq!(dest[3], 0x02);
+        assert_eq!(dest[5], 0x11);
+
+        // Length in 32bits (4 bytes) + Contents (5 elements)
+        assert_eq!(value.write_size(0), dest.len());
+    }
+
+    #[test]
+    fn test_bytebuf_encodes_transparent_with_vecu8() {
+        let raw_data: [u8; 10] = [0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, 0x09, 0x0A];
+
+        // Data encoded for Vec<u8> will have the first 4 bytes (32 bits) for
+        // data length, and the remaining bytes represents the actual data
+        let encoded_expect: [u8; 14] = [
+            0x00, 0x00, 0x00, 0x0A, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, 0x09, 0x0A,
+        ];
+
+        let mut encoded_data: Vec<u8> = Vec::default();
+        let encoded_res = Vec::<u8>::from(raw_data).encode(&mut encoded_data, 0);
+
+        assert!(encoded_res.is_ok());
+        assert_eq!(
+            encoded_expect,
+            encoded_data.as_slice(),
+            "encoded version doesn't match with expected"
+        );
+        assert_eq!(encoded_data[3], 0x0A, "the length value doesnt match");
+
+        let mut bytebuf_encoded = ByteBuf::default();
+        let encode_bytebuf_res = ByteBuf::from(Vec::from(raw_data)).encode(&mut bytebuf_encoded, 0);
+
+        assert!(encode_bytebuf_res.is_ok());
+        assert_eq!(
+            bytebuf_encoded.inner,
+            encoded_expect.as_slice(),
+            "encoded version doesn't match with expected"
+        );
+        assert_eq!(
+            bytebuf_encoded.inner,
+            encoded_data.as_slice(),
+            "encoded version doesn't match with expected"
+        );
+        assert_eq!(
+            bytebuf_encoded.inner[3], 0x0A,
+            "the length value doesnt match"
+        );
+    }
+
+    #[test]
+    fn test_bytebuf_decode_transparent_with_vecu8() {
```

Comment:
This will make sure the output is the same as for `Vec<u8>`. Which will ensure the output is compliant with what we use today. Also applies for the test above on: `test_bytebuf_encodes_transparent_with_vecu8`